Fix interrupt handler for endpoints > 2

mjrgh opened the pull request:
The IRQ handler calculated the endpoint number wrong when handling IN and OUT tokens.  The errant code worked by accident for endpoints 1 and 2 but was incorrect for other addresses.  For endpoints 5 and above it resulted in stray pointer read/writes that could crash the device.

0xc0170 commented:
Can you please sign : https://developer.mbed.org/contributor_agreement/ ?

mjrgh commented:
Hi - I already signed that a couple of years ago.
Regards,Mike

Date: Mon, 4 Jan 2016 01:34:11 -0800
From: notifications@github.com
To: mbed@noreply.github.com
CC: mjr_@hotmail.com
Subject: Re: [mbed] Fix interrupt handler for endpoints > 2 (#1481)

Can you please sign : https://developer.mbed.org/contributor_agreement/ ?


—
Reply to this email directly or view it on GitHub.

0xc0170 commented:
@mjrgh I could not locate username mjrgh in mbed, please share it to verify

mjrgh commented:
Hmm... strange.  When I go to the contributor agreement page, it doesn't offer an option to sign, it just says "You accepted the Contributor Agreement on Tue 05 Aug 2014".  
 >please share it to verify
Not sure what you mean - could you be more specific?
Regards,Mike
Date: Mon, 11 Jan 2016 02:32:35 -0800
From: notifications@github.com
To: mbed@noreply.github.com
CC: mjr_@hotmail.com
Subject: Re: [mbed] Fix interrupt handler for endpoints > 2 (#1481)

@mjrgh I could not locate username mjrgh in mbed, please share it to verify


—
Reply to this email directly or view it on GitHub."

mjrgh commented:
Oh, by the way, I think my previous changes to the same code were pushed under my mbed user ID, "mjr" - the new ones are on my github account, which might be why they don't appear to be connected.
Regards,Mike

From: mjr_@hotmail.com
To: reply@reply.github.com
Subject: RE: [mbed] Fix interrupt handler for endpoints > 2 (#1481)
Date: Mon, 11 Jan 2016 20:46:55 +0000




Hmm... strange.  When I go to the contributor agreement page, it doesn't offer an option to sign, it just says "You accepted the Contributor Agreement on Tue 05 Aug 2014".  
 >please share it to verify
Not sure what you mean - could you be more specific?
Regards,Mike
Date: Mon, 11 Jan 2016 02:32:35 -0800
From: notifications@github.com
To: mbed@noreply.github.com
CC: mjr_@hotmail.com
Subject: Re: [mbed] Fix interrupt handler for endpoints > 2 (#1481)

@mjrgh I could not locate username mjrgh in mbed, please share it to verify


—
Reply to this email directly or view it on GitHub."

0xc0170 commented:
Thanks, for sharing your mbed username. And for these fixes, I noticed you got some other fixes on mbed in your fork :+1:

mjrgh commented:
Great - thanks!
Regards,Mike

Date: Thu, 14 Jan 2016 23:44:14 -0800
From: notifications@github.com
To: mbed@noreply.github.com
CC: mjr_@hotmail.com
Subject: Re: [mbed] Fix interrupt handler for endpoints > 2 (#1481)

Thanks, for sharing your mbed username. And for these fixes, I noticed you got some other fixes on mbed in your fork  


—
Reply to this email directly or view it on GitHub.

mjrgh commented:
By the way, I've done a more extensive overhaul of the KL25Z USB HAL implementation, which I haven't entered a pull request for, because the changes are just too big.  I'm working on a project that involves multiple endpoints (that's how I ran into the bugs that these past couple of pull requests were about), and that exposed a bunch of race conditions in the KL25Z HAL code.  The race conditions mostly manifest as random USB disconnects.  Most of the race condition problems are apparent from a careful reading of the code, and they'll affect all existing users of the code, but the probability seems low enough for most simple applications that I guess no one has noticed them until now.
The race conditions strike randomly, of course, and they seem to happen quite rarely in simple code with one non-control endpoint.  The previous version of my project used a simple USB joystick interface with one non-control endpoint, and the connection was *mostly* stable.  I did find it necessary to put in 1ms waits around all of my USB read/write operations, which I previously chalked up to hardware bugs, but I now realize the race conditions were the real culprit.  The 1ms waits helped simply because they slowed things down and thereby reduced the probability of a USB interrupt coming in while the read/write call was in progress.  
I think I've made everything deterministic and properly serialized in the overhaul.  I added critical sections around all (hopefully!) of the code that touches shared statics and that can run from application context.  I also changed the way endpoint un-stall is handled - those didn't work reliably in the official version, so a stall would generally kill a connection even though it shouldn't have.
My test case seems rock-solid with the changes.  I was able to trigger disconnects within a matter of minutes with the original code, but the new version runs for days at a time without every disconnecting.
If you're interested in taking a look, I published my fork here:
http://mbed.org/users/mjr/code/USBDevice/
I'm sure the changes are too extensive to just drop into the official library, but I thought I'd share it in case you or other of the platform developers want to take a look at it and consider adapting it.  And in any case, the fork is there in case anyone has similar problems and wants to try this version to see if it helps.
Regards,Mike

From: mjr_@hotmail.com
To: reply@reply.github.com
Subject: RE: [mbed] Fix interrupt handler for endpoints > 2 (#1481)
Date: Fri, 15 Jan 2016 18:44:14 +0000




Great - thanks!
Regards,Mike

Date: Thu, 14 Jan 2016 23:44:14 -0800
From: notifications@github.com
To: mbed@noreply.github.com
CC: mjr_@hotmail.com
Subject: Re: [mbed] Fix interrupt handler for endpoints > 2 (#1481)

Thanks, for sharing your mbed username. And for these fixes, I noticed you got some other fixes on mbed in your fork  


—
Reply to this email directly or view it on GitHub.

0xc0170 commented:
@mjrgh Users will benefit from those bugfixes in the upstream. If you can pack those changes as patches with bug fixes, as this one, would be nice. Thanks for sharing all this devel

cc @sg-
